### PR TITLE
nixos/paperless: override enabled tesseract languages with the in paperless configured ones

### PR DESCRIPTION
**Pull request by SuperSandro2000**
## Description of changes

Reduces required closure size dramatically when only a few languages are required.

```
all: ε → ∅, -1039163.4 KiB
deu.traineddata: ∅ → ε, +15075.8 KiB
eng.traineddata: ∅ → ε, +22916.8 KiB
equ.traineddata: ∅ → ε, +2199.3 KiB
osd.traineddata: ∅ → ε, +10315.4 KiB
tesseract: -31.5 KiB
```

## Things done

Tested with and without PAPERLESS_OCR_LANGUAGE being configured.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by leona-ya — APPROVED** [review]
This seems quite nice. Still, I would like someone else to also have a look at it.

File: pkgs/applications/office/paperless-ngx/default.nix [hunk 1]
```diff
@@ -37,17 +39,21 @@ let
...
+        uvicorn = prev.uvicorn.overridePythonAttrs (_: {
+          version = "0.25.0";
+          src = fetchFromGitHub {
+            owner = "encode";
+            repo = "uvicorn";
+            rev = "0.25.0";
+            hash = "sha256-ng98DTw49zyFjrPnEwfnPfONyjKKZYuLl0qduxSppYk=";
+          };
+        });
+      })
```
Comment (ambroisie) [inline pkgs/applications/office/paperless-ngx/default.nix:55]: Should we perhaps switch the order, _just in case_ someone wants to override `uvicorn`?

**Review by ambroisie — APPROVED** [review]
LGTM.

**Review by erikarvstedt — CHANGES_REQUESTED** [review]
Why not simply override `ocrmypdf` in the paperless pkg, [like so](https://github.com/erikarvstedt/nixpkgs/commit/3d804ec78ce86a91e7d96410cc65d58c88100104)?

Maybe keep the `pkg = cfg.package`  variable. It shortens some complex expressions, also the diff from this PR gets much smaller. Medium preference.

**Comment by leona-ya** [comment]
> Why not simply override `ocrmypdf` in the paperless pkg, [like so](https://github.com/erikarvstedt/nixpkgs/commit/3d804ec78ce86a91e7d96410cc65d58c88100104)?

This should not produce the same result, as in Sandro's version `ocrmypdf.override { tesseract = tesseract5; }` is using the tesseract defined above in the rec clause, while your version just uses the normal input

**Comment by erikarvstedt** [comment]
No, my version is equivalent.
The `tesseract5` input in the `paperless-ngx` pkg is exactly the customized tesseract pkg from the rec attrs.

**Comment by leona-ya** [comment]
Oh my bad, one more override than I thought about, sorry :/

**Comment by SuperSandro2000** [comment]
Moved the overwrite like in the linked patch. Thanks!

**Review by leona-ya — APPROVED** [review]

File: nixos/modules/services/misc/paperless.nix [hunk 2]
```diff
@@ -200,7 +199,17 @@ in
...
+      apply = pkg: pkg.override {
+        tesseract5 = pkg.tesseract5.override {
+          # always enable detection modules
+          enableLanguages = if cfg.settings ? PAPERLESS_OCR_LANGUAGE then
+            [ "equ" "osd" ]
+              ++ lib.splitString "+" cfg.settings.PAPERLESS_OCR_LANGUAGE
+          else null;
+        };
+      };
+    };
```
Comment (mweinelt) [inline nixos/modules/services/misc/paperless.nix:212]: This breaks overridePythonAttrs used on the package.

Comment (SuperSandro2000) [reply nixos/modules/services/misc/paperless.nix:212]: What should then be used instead? When I last looked at this, it wasn't easily possible to do an override in the package for the python package set and here.

Comment (mweinelt) [reply nixos/modules/services/misc/paperless.nix:212]: It could be an override on the default package, that users need to carry over.

Comment (SuperSandro2000) [reply nixos/modules/services/misc/paperless.nix:212]: Since the override is a bit more complicated, I am not sure that this is the best option.

**Pull request merged**